### PR TITLE
feat: Support for ModelBuilder In_Process Mode (1/2)

### PR DESCRIPTION
**Pull request by bryannahm1**
*Issue #, if available:*

*Description of changes:*
Adding new mode of deployment named in_process_mode. Added in_process script to mode, made edits to transformers and model builder in order to support.

*Testing done:*
Integr tests run locally.

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [X] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [X] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [X] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [X] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [X] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [X] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/serve/builder/model_builder.py [hunk 1]
```diff
@@ -896,8 +913,11 @@ def build(  # pylint: disable=R0911
 
     def _build_validations(self):
         """Validations needed for model server overrides, or auto-detection or fallback"""
-        if self.mode == Mode.IN_PROCESS:
-            raise ValueError("IN_PROCESS mode is not supported yet!")
+        if self.mode == Mode.IN_PROCESS and self.model_server is not ModelServer.MMS:
+            raise ValueError(
+                "IN_PROCESS mode is not supported yet for model server. It is "
+                "supported for MMS/Transformers server in beta release."
```
Comment (makungaj1) [inline src/sagemaker/serve/builder/model_builder.py:919]: nit: `IN_PROCESS mode is only supported for MMS/Transformers server in beta release.`

File: src/sagemaker/serve/builder/transformers_builder.py [hunk 2]
```diff
@@ -161,7 +164,7 @@ def _get_hf_metadata_create_model(self) -> Type[Model]:
                 vpc_config=self.vpc_config,
             )
 
-        if not self.image_uri and self.mode == Mode.LOCAL_CONTAINER:
+        if self.mode == Mode.LOCAL_CONTAINER or self.mode == Mode.IN_PROCESS:
```
Comment (makungaj1) [inline src/sagemaker/serve/builder/transformers_builder.py:167]: nit:
```
LOCAL_MODES = [Mode.LOCAL_CONTAINER, Mode.IN_PROCESS]
if self.mode in LOCAL_MODES:
      ...
```

File: src/sagemaker/serve/builder/transformers_builder.py [hunk 3]
```diff
@@ -274,7 +293,7 @@ def _build_transformers_env(self):
 
         self.pysdk_model = self._create_transformers_model()
 
-        if self.mode == Mode.LOCAL_CONTAINER:
+        if self.mode == Mode.LOCAL_CONTAINER or self.mode == Mode.IN_PROCESS:
```
Comment (makungaj1) [inline src/sagemaker/serve/builder/transformers_builder.py:296]: nit:
```
if self.mode in LOCAL_MODES:
      ...
```

File: src/sagemaker/serve/mode/in_process_mode.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+from sagemaker.serve.utils.exceptions import LocalDeepPingException
+from sagemaker.serve.model_server.multi_model_server.server import InProcessMultiModelServer
+from sagemaker.session import Session
+
+logger = logging.getLogger(__name__)
+
+_PING_HEALTH_CHECK_INTERVAL_SEC = 5
+
+_PING_HEALTH_CHECK_FAIL_MSG = (
+    "Container did not pass the ping health check. "
```
Comment (makungaj1) [inline src/sagemaker/serve/mode/in_process_mode.py:22]: Does IN_PROCESS mode uses `Container`?

File: src/sagemaker/serve/mode/in_process_mode.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        self,
+        image: str,
+        secret_key: str,
+        predictor: PredictorBase,
+        env_vars: Dict[str, str] = None,
+        model_path: str = None,
+    ):
+        """Placeholder docstring"""
+
+        # self._pull_image(image=image)
```
Comment (samruds) [inline src/sagemaker/serve/mode/in_process_mode.py:80]: Can remove this this comment

File: src/sagemaker/serve/model_server/multi_model_server/server.py [hunk 6]
```diff
@@ -20,6 +20,83 @@
...
+            return response.content
+        except Exception as e:
+            raise Exception("Unable to send request to the local container server") from e
+
+        return (True, response)
+
+    def _multi_model_server_deep_ping(self, predictor: PredictorBase):
+        """Placeholder docstring"""
+        response = None
+        logger.debug("AM I HERE? PING PING")
```
Comment (samruds) [inline src/sagemaker/serve/model_server/multi_model_server/server.py:86]: Remove this line

File: src/sagemaker/serve/model_server/multi_model_server/server.py [hunk 7]
```diff
@@ -20,6 +20,83 @@
...
+                Path(model_path).joinpath("code"): {
+                    "bind": MODE_DIR_BINDING,
+                    "mode": "rw",
+                },
+            },
+            environment=env_vars,
+        )
+
+    def _invoke_multi_model_server_serving(self, request: object, content_type: str, accept: str):
+        """Placeholder docstring"""
```
Comment (samruds) [inline src/sagemaker/serve/model_server/multi_model_server/server.py:62]: Update doc strings

File: tests/unit/sagemaker/serve/builder/test_model_builder.py [hunk 8]
```diff
@@ -67,10 +67,18 @@
 class TestModelBuilder(unittest.TestCase):
     @patch("sagemaker.serve.builder.model_builder._ServeSettings")
     def test_validation_in_progress_mode_not_supported(self, mock_serveSettings):
```
Comment (samruds) [inline tests/unit/sagemaker/serve/builder/test_model_builder.py:69]: Can we add a test a where it is supported?

File: src/sagemaker/serve/builder/model_builder.py (hunk 1 above)
Comment (bryannahm1) [reply src/sagemaker/serve/builder/model_builder.py:919]: This is better wording, I will change it thank you!

File: src/sagemaker/serve/builder/transformers_builder.py (hunk 2 above)
Comment (bryannahm1) [reply src/sagemaker/serve/builder/transformers_builder.py:167]: Good suggestion, I've made the edit, thank you

File: src/sagemaker/serve/builder/transformers_builder.py (hunk 3 above)
Comment (bryannahm1) [reply src/sagemaker/serve/builder/transformers_builder.py:296]: Made the change

File: src/sagemaker/serve/mode/in_process_mode.py (hunk 4 above)
Comment (bryannahm1) [reply src/sagemaker/serve/mode/in_process_mode.py:22]: No it does not, I will be sure to change this, good catch.

File: src/sagemaker/serve/mode/in_process_mode.py (hunk 5 above)
Comment (bryannahm1) [reply src/sagemaker/serve/mode/in_process_mode.py:80]: Removed

File: src/sagemaker/serve/model_server/multi_model_server/server.py (hunk 6 above)
Comment (bryannahm1) [reply src/sagemaker/serve/model_server/multi_model_server/server.py:86]: Removed it, thank you

File: src/sagemaker/serve/model_server/multi_model_server/server.py (hunk 7 above)
Comment (bryannahm1) [reply src/sagemaker/serve/model_server/multi_model_server/server.py:62]: The docstrings are updated, good catch.

File: tests/unit/sagemaker/serve/builder/test_model_builder.py (hunk 8 above)
Comment (bryannahm1) [reply tests/unit/sagemaker/serve/builder/test_model_builder.py:69]: Yes, I will add it. Good suggestion.

File: src/sagemaker/serve/model_server/multi_model_server/server.py [hunk 9]
```diff
@@ -20,6 +20,82 @@
...
+
+            logger.info(response.content)
+
+            return response.content
+        except Exception as e:
+            raise Exception("Unable to send request to the local container server") from e
+
+        return (True, response)
+
+    def _multi_model_server_deep_ping(self, predictor: PredictorBase):
```
Comment (makungaj1) [inline src/sagemaker/serve/model_server/multi_model_server/server.py:83]: is this complete?

File: src/sagemaker/serve/model_server/multi_model_server/server.py [hunk 10]
```diff
@@ -20,6 +20,82 @@
...
+            "SAGEMAKER_PROGRAM": "inference.py",
+            "SAGEMAKER_SERVE_SECRET_KEY": secret_key,
+            "LOCAL_PYTHON": platform.python_version(),
+        }
+        if env_vars:
+            env_vars.update(env)
+        else:
+            env_vars = env
+
+        self.container = client.containers.run(
```
Comment (makungaj1) [inline src/sagemaker/serve/model_server/multi_model_server/server.py:46]: Are we spinning up a docker container or using fast api for serving?

Comment (bryannahm1) [reply src/sagemaker/serve/model_server/multi_model_server/server.py:46]: The container will be stubbed, this is only 1/2 of the full implementation of InProcess mode. My next PR will include the FastAPI.

File: src/sagemaker/serve/model_server/multi_model_server/server.py (hunk 9 above)
Comment (bryannahm1) [reply src/sagemaker/serve/model_server/multi_model_server/server.py:83]: I have stubbed it.

File: src/sagemaker/serve/utils/exceptions.py [hunk 11]
```diff
@@ -1,4 +1,4 @@
-"""Placeholder Docstring"""
+"""Exceptions used across different model builder invocations"""
```
[the remark this answers is not in the document]
Comment (bryannahm1) [reply src/sagemaker/serve/utils/exceptions.py:1]: Of course, thanks for the suggestion.

File: src/sagemaker/serve/model_server/multi_model_server/server.py [hunk 12]
```diff
@@ -59,7 +135,7 @@ def _start_serving(
         )
 
     def _invoke_multi_model_server_serving(self, request: object, content_type: str, accept: str):
-        """Placeholder docstring"""
+        """Invokes MMS server by hitting the docker host"""
```
[the remark this answers is not in the document]
Comment (bryannahm1) [reply src/sagemaker/serve/model_server/multi_model_server/server.py:138]: Thanks !

File: src/sagemaker/serve/mode/in_process_mode.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+    ):
+        # pylint: disable=bad-super-call
+        super().__init__()
+
+        self.inference_spec = inference_spec
+        self.model_path = model_path
+        self.env_vars = env_vars
+        self.session = session
+        self.schema_builder = schema_builder
+        self.ecr = session.boto_session.client("ecr")
```
[the remark this answers is not in the document]
Comment (bryannahm1) [reply src/sagemaker/serve/mode/in_process_mode.py:47]: Removed.

File: src/sagemaker/serve/builder/transformers_builder.py [hunk 14]
```diff
@@ -161,7 +165,7 @@ def _get_hf_metadata_create_model(self) -> Type[Model]:
                 vpc_config=self.vpc_config,
             )
 
-        if not self.image_uri and self.mode == Mode.LOCAL_CONTAINER:
+        if self.mode in LOCAL_MODES:
```
[the remark this answers is not in the document]
Comment (bryannahm1) [reply src/sagemaker/serve/builder/transformers_builder.py:168]: Great catch, I had missed that. It is returned to how it was

File: src/sagemaker/serve/model_server/multi_model_server/server.py [hunk 15]
```diff
@@ -20,6 +20,82 @@
...
+            volumes={
+                Path(model_path).joinpath("code"): {
+                    "bind": MODE_DIR_BINDING,
+                    "mode": "rw",
+                },
+            },
+            environment=env_vars,
+        )
+
+    def _invoke_multi_model_server_serving(self, request: object, content_type: str, accept: str):
```
[the remark this answers is not in the document]
Comment (bryannahm1) [reply src/sagemaker/serve/model_server/multi_model_server/server.py:61]: I have stubbed it, thank you.

File: src/sagemaker/serve/mode/in_process_mode.py [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+
+        if self.model_server == ModelServer.MMS:
+            self._start_serving(
+                client=self.client,
+                image=image,
+                model_path=model_path if model_path else self.model_path,
+                secret_key=secret_key,
+                env_vars=env_vars if env_vars else self.env_vars,
+            )
+            logger.info("Starting PING")
```
[the remark this answers is not in the document]
Comment (bryannahm1) [reply src/sagemaker/serve/mode/in_process_mode.py:90]: Removed :)

File: src/sagemaker/serve/builder/model_builder.py [hunk 17]
```diff
@@ -896,8 +913,15 @@ def build(  # pylint: disable=R0911
...
     def _build_validations(self):
         """Validations needed for model server overrides, or auto-detection or fallback"""
-        if self.mode == Mode.IN_PROCESS:
-            raise ValueError("IN_PROCESS mode is not supported yet!")
+        if self.mode == Mode.IN_PROCESS and self.model_server is not ModelServer.MMS:
+            raise ValueError(
+                "IN_PROCESS mode is only supported for MMS/Transformers server in beta release."
+            )
+
+        if self.mode == Mode.IN_PROCESS and self.model_server == ModelServer.MMS:
```
[the remark this answers is not in the document]
Comment (bryannahm1) [reply src/sagemaker/serve/builder/model_builder.py:921]: I've updated it, thank you, let me know if there's anything else I can add.

**Review by samruds — APPROVED** [review]
LGTM, please ensure existing servers dont break. Please run notebook against this change to ensure backward compatibility...

**Comment by makungaj1** [comment]
Can we have this merge in a feature branch and merge to master until all complete and fully tested? cc @samruds

**Comment by samruds** [comment]
> Can we have this merge in a feature branch and merge to master until all complete and fully tested? cc @samruds

Proposed split to enable to merge to mainline. 

1/2 -> introduce in process mode , it should hit no-op stubs and return an empty response. Other servers should not be impacted.

2/2 - we introduce the fast API logic into the stubs. At this point the response should be a valid inference response. Other servers should not be impacted.

**Review by makungaj1 — APPROVED** [review]

**Review by samruds — APPROVED** [review]

**Review by makungaj1 — APPROVED** [review]

**Review by sage-maker — APPROVED** [review]

**Pull request merged**